flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java

```diff
@@ -159,17 +168,51 @@ void applyRemovals() {
 
     @Override
     public void notifyOfAddedMetric(Metric metric, String metricName, MetricGroup group) {
-        MetricGroup metricGroup = unwrap(group);
-        LOG.debug("Registered {} @ {}", metricName, metricGroup);
-        synchronized (this) {
-            metrics.computeIfAbsent(metricGroup, dummy -> new HashMap<>()).put(metricName, metric);
+        if (captureMode != CaptureMode.NONE) {
+            MetricGroup metricGroup = unwrap(group);
+            LOG.debug("Registered {} @ {}", metricName, metricGroup);
+            synchronized (this) {
+                metrics.computeIfAbsent(metricGroup, dummy -> new HashMap<>())
```

Comment:
WDYT about making metrics a `ConcurrentHashMap`? It seems to me that we mostly only synchronize on the object to prevent concurrent access to the metrics map.

Reply:
I'm always having a hard time to reason about nested data structures and their synchronization behavior. For example, could a metric be lost because we eagerly deregister the last metric at the same time a new metric is added? Then `notifyOfRemovedMetric` might fetch the group first and determines it is empty. At the same time, `notifyOfAddedMetric` has fetched the still-registered but empty group and added the new metric. But this is now lost since `notifyOfRemovedMetric` eventually removes it.
So before trying to optimize it, I first would go with the simplest approach. If we see it becomes a bottle-neck (which I doubt) we can still optimize it later. Then I might have the mental capacity for thinking all cases through (or we find someone else).

Reply:
I am also not concerned about the performance but the readability and maintainability. To me having dedicated synchronized blocks already implies that the coming code section is somewhat critical and not thread-safe but seeing it only doing a single operation feels strange. 
Moreover, I think the current threading model is more complicated because you must go through the complete class and have a look at all `synchronized` blocks before you can reason whether it is solid or not. By moving this responsibility to proper java builtins the contract is immediately clear when seeing the variable declarations.

Reply:
I think one easy thing is probably to `synchronize` the two methods `notifyOfRemovedMetric` `notifyOfAddedMetric` if performance is no concern here and will add at least some clarity.